File: src/portfolio_manager/services/price_service.py
"""Service for fetching stock prices."""

from calendar import monthrange
from datetime import date, timedelta
from decimal import Decimal, InvalidOperation

import httpx


class PriceService:
    """Service for fetching stock prices."""

    def __init__(
        self,
        price_client,
        price_cache_repository=None,
        today_provider=date.today,
    ):
        """Initialize with a price client."""
        self.price_client = price_client
        self.price_cache_repository = price_cache_repository
        self.today_provider = today_provider
        self._price_cache: dict[
            tuple[str, str | None], tuple[Decimal, str, str, str | None]
        ] = {}
        self._change_rates_cache: dict[
            tuple[str, date, str | None], dict[str, Decimal]
        ] = {}

    def get_stock_price(
        self, ticker: str, preferred_exchange: str | None = None
    ) -> tuple[Decimal, str, str, str | None]:
        """Get current price, currency, name, and exchange for a stock ticker."""
        # Check in-memory cache first (keyed by ticker and exchange)
        cache_key = (ticker, preferred_exchange)
        if cache_key in self._price_cache:
            return self._price_cache[cache_key]

        today = self.today_provider()
        cached = None
        if self.price_cache_repository:
            cached = self.price_cache_repository.get_by_ticker_and_date(ticker, today)
        if cached and cached.price > 0:
            result = (
                cached.price,
                cached.currency,
                cached.name,
                cached.exchange,
            )
            self._price_cache[cache_key] = result
            return result

        quote = self.price_client.get_price(
            ticker, preferred_exchange=preferred_exchange
        )
        price = Decimal(str(quote.price))
        if self.price_cache_repository and price > 0:
            self.price_cache_repository.save(
                ticker=ticker,
                price_date=today,
                price=price,
                currency=quote.currency,
                name=quote.name,
                exchange=quote.exchange,
            )
        result = (price, quote.currency, quote.name, quote.exchange)
        # Only cache valid prices (price > 0)
        if price > 0:
            self._price_cache[cache_key] = result
        return result

    def get_stock_change_rates(
        self,
        ticker: str,
        as_of: date | None = None,
        preferred_exchange: str | None = None,
    ) -> dict[str, Decimal]:
        """Get 1Y/6M/1M change rates compared to historical close prices."""
        if as_of is None:
            as_of = date.today()

        # Check in-memory cache first (keyed by ticker, date, and exchange)
        cache_key = (ticker, as_of, preferred_exchange)
        if cache_key in self._change_rates_cache:
            return self._change_rates_cache[cache_key]

        def shift_years(base_date: date, years: int) -> date:
            target_year = base_date.year - years
            last_day = monthrange(target_year, base_date.month)[1]
            target_day = min(base_date.day, last_day)
            return date(target_year, base_date.month, target_day)

        def shift_months(base_date: date, months: int) -> date:
            target_year = base_date.year
            target_month = base_date.month - months
            while target_month <= 0:
                target_month += 12
                target_year -= 1
            last_day = monthrange(target_year, target_month)[1]
            target_day = min(base_date.day, last_day)
            return date(target_year, target_month, target_day)

        def adjust_to_previous_business_day(target_date: date) -> date:
            if target_date.weekday() == 5:
                return target_date - timedelta(days=1)
            if target_date.weekday() == 6:
                return target_date - timedelta(days=2)
            return target_date

        current_price, currency, name, exchange = self.get_stock_price(
            ticker, preferred_exchange=preferred_exchange
        )
        targets = {
            "1y": adjust_to_previous_business_day(shift_years(as_of, 1)),
            "6m": adjust_to_previous_business_day(shift_months(as_of, 6)),
            "1m": adjust_to_previous_business_day(shift_months(as_of, 1)),
        }
        change_rates: dict[str, Decimal] = {}
        for label, target_date in targets.items():
            cached = None
            if self.price_cache_repository:
                cached = self.price_cache_repository.get_by_ticker_and_date(
                    ticker, target_date
                )
            if cached and cached.price > 0:
                past_close = cached.price
            else:
                try:
                    past_close = Decimal(
                        str(
                            self.price_client.get_historical_close(
                                ticker,
                                target_date,
                                preferred_exchange=preferred_exchange,
                            )
                        )
                    )
                except (httpx.HTTPStatusError, InvalidOperation, ValueError, TypeError):
                    past_close = Decimal("0")
                if self.price_cache_repository and past_close > 0:
                    self.price_cache_repository.save(
                        ticker=ticker,
                        price_date=target_date,
                        price=past_close,
                        currency=currency,
                        name=name,
                        exchange=exchange,
                    )
            if past_close == 0:
                change_rates[label] = Decimal("0")
            else:
                change_rates[label] = (
                    (current_price - past_close) / past_close * Decimal("100")
                )
        self._change_rates_cache[cache_key] = change_rates
        return change_rates

File: internal/services/price_service.go
package services

import (
	"context"
	"time"

	"github.com/kadragon/portfolio-manager/internal/datex"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/numeric"
	"github.com/kadragon/portfolio-manager/internal/repositories"
)

// supportedPeriods is the set of valid change-rate period labels.
var supportedPeriods = map[string]bool{"1d": true, "1m": true, "6m": true, "1y": true}

// orderToPrice maps order-form exchange codes to price-form codes (KIS convention).
var orderToPrice = map[string]string{
	"NAS": "NASD",
	"NYS": "NYSE",
	"AMS": "AMEX",
}

// priceToOrder is the reverse mapping.
var priceToOrder = map[string]string{
	"NASD": "NAS",
	"NYSE": "NYS",
	"AMEX": "AMS",
}

// PriceService resolves stock prices from the DB cache only.
// PriceSyncService owns live API access and keeps the DB up to date.
type PriceService struct {
	stockPrices   *repositories.StockPriceRepository
	todayProvider func() time.Time
}

// NewPriceService creates a DB-only PriceService. Use PriceSyncService for live fetching.
func NewPriceService(stockPrices *repositories.StockPriceRepository) *PriceService {
	return &PriceService{
		stockPrices:   stockPrices,
		todayProvider: func() time.Time { return ktime.Now().Time },
	}
}

// WithTodayProvider overrides the service's date source. Use in tests to fix today's date.
func (s *PriceService) WithTodayProvider(fn func() time.Time) *PriceService {
	s.todayProvider = fn
	return s
}

// GetStockPrice returns (price, currency, name, exchange) for ticker from DB.
// Returns today's price if available, otherwise the most recent cached price.
// Returns zero when no data exists.
func (s *PriceService) GetStockPrice(ctx context.Context, ticker, preferredExchange string) (numeric.Decimal, string, string, string) {
	orderExch := toOrderExchange(preferredExchange)
	today := datex.FromTime(s.todayProvider())
	if sp := s.loadCached(ctx, ticker, today); sp != nil {
		return sp.Price, sp.Currency, sp.Name, toOrderExchange(sp.Exchange.String)
	}
	return numeric.Zero, "KRW", "", orderExch
}

// GetCachedPrice returns the stored price for (ticker, date) from the DB, or nil.
func (s *PriceService) GetCachedPrice(ctx context.Context, ticker string, date datex.Date) *models.StockPrice {
	return s.loadCached(ctx, ticker, date)
}

// loadCached returns today's price or, if absent, the most-recent stored price (stale fallback).
// Use getExact when the fallback would give misleading results (e.g. change-rate history).
func (s *PriceService) loadCached(ctx context.Context, ticker string, date datex.Date) *models.StockPrice {
	if s.stockPrices == nil {
		return nil
	}
	sp, _ := s.stockPrices.GetByTickerAndDate(ctx, ticker, date)
	if sp != nil && sp.Price.IsPositive() {
		return sp
	}
	// Fall back to most-recent available price (handles weekends and holidays).
	sp, _ = s.stockPrices.GetLatestByTicker(ctx, ticker)
	if sp != nil && sp.Price.IsPositive() {
		return sp
	}
	return nil
}

// getOnOrBefore returns the most recent stored price at or before the given date,
// or nil. Unlike loadCached it never jumps forward to the latest price, so a
// non-business target date resolves to the nearest prior trading day rather than
// today's value — correct for historical change-rate lookups.
func (s *PriceService) getOnOrBefore(ctx context.Context, ticker string, date datex.Date) *models.StockPrice {
	if s.stockPrices == nil {
		return nil
	}
	sp, _ := s.stockPrices.GetOnOrBeforeDate(ctx, ticker, date)
	if sp != nil && sp.Price.IsPositive() {
		return sp
	}
	return nil
}

// GetStockChangeRates returns rate-of-change (%) for each period from DB.
// Returns nil when no current price is available or no valid periods requested.
func (s *PriceService) GetStockChangeRates(ctx context.Context, ticker, preferredExchange string, periods []string) map[string]numeric.Decimal {
	if s.stockPrices == nil {
		return nil
	}

	normalized := make([]string, 0, len(periods))
	seen := map[string]bool{}
	for _, p := range periods {
		if supportedPeriods[p] && !seen[p] {
			normalized = append(normalized, p)
			seen[p] = true
		}
	}
	if len(normalized) == 0 {
		return nil
	}

	currentPrice, _, _, _ := s.GetStockPrice(ctx, ticker, preferredExchange)
	if currentPrice.IsZero() {
		return nil
	}

	today := datex.FromTime(s.todayProvider())
	targetDates := computeTargetDates(today.Time)
	result := make(map[string]numeric.Decimal, len(normalized))

	for _, label := range normalized {
		target := targetDates[label]
		targetDate := datex.FromTime(target)

		var pastClose numeric.Decimal
		if cached := s.getOnOrBefore(ctx, ticker, targetDate); cached != nil && cached.Price.IsPositive() {
			pastClose = cached.Price
		}

		if pastClose.IsZero() {
			result[label] = numeric.Zero
			continue
		}
		rate := numeric.Wrap(
			currentPrice.Sub(pastClose.Decimal).Div(pastClose.Decimal).Mul(hundred.Decimal),
		)
		result[label] = rate
	}
	return result
}

// GetStockChangeSince returns rate-of-change (%) from the nearest cached close
// at or before startDate to the current cached price.
func (s *PriceService) GetStockChangeSince(ctx context.Context, ticker string, startDate datex.Date) *numeric.Decimal {
	if s.stockPrices == nil || startDate.Time.IsZero() {
		return nil
	}
	currentPrice, _, _, _ := s.GetStockPrice(ctx, ticker, "")
	if currentPrice.IsZero() {
		return nil
	}
	start := s.getOnOrBefore(ctx, ticker, startDate)
	if start == nil || !start.Price.IsPositive() {
		return nil
	}
	rate := numeric.Wrap(
		currentPrice.Sub(start.Price.Decimal).Div(start.Price.Decimal).Mul(hundred.Decimal),
	)
	return &rate
}

// computeTargetDates builds the historical reference date for each period.
func computeTargetDates(today time.Time) map[string]time.Time {
	return map[string]time.Time{
		"1y": prevBizDay(shiftYears(today, 1)),
		"6m": prevBizDay(shiftMonths(today, 6)),
		"1m": prevBizDay(shiftMonths(today, 1)),
		"1d": prevBizDay(today.AddDate(0, 0, -1)),
	}
}

func shiftYears(t time.Time, years int) time.Time {
	target := t.AddDate(-years, 0, 0)
	y, m, _ := target.Date()
	last := lastDayOfMonth(y, m)
	d := t.Day()
	if d > last {
		d = last
	}
	return time.Date(y, m, d, 0, 0, 0, 0, target.Location())
}

func shiftMonths(t time.Time, months int) time.Time {
	target := t.AddDate(0, -months, 0)
	y, m, _ := target.Date()
	last := lastDayOfMonth(y, m)
	d := t.Day()
	if d > last {
		d = last
	}
	return time.Date(y, m, d, 0, 0, 0, 0, target.Location())
}

func lastDayOfMonth(year int, month time.Month) int {
	return time.Date(year, month+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

func prevBizDay(t time.Time) time.Time {
	switch t.Weekday() {
	case time.Saturday:
		return t.AddDate(0, 0, -1)
	case time.Sunday:
		return t.AddDate(0, 0, -2)
	default:
		return t
	}
}

func toOrderExchange(e string) string {
	if v, ok := priceToOrder[e]; ok {
		return v
	}
	return e
}

func toPriceExchange(e string) string {
	if v, ok := orderToPrice[e]; ok {
		return v
	}
	return e
}
